cli/src/cmd/forge/script/runner.rs

```diff
@@ -0,0 +1,222 @@
+                    )
+                }
+                Err(e) => return Err(e.into()),
+            }
+        };
+
+        Ok((
+            address,
+            ScriptResult {
+                returned: bytes::Bytes::new(),
+                success,
+                gas,
+                labeled_addresses,
+                transactions,
+                logs,
+                traces,
+                debug,
+            },
+        ))
+    }
+
+    /// Executes the method that will collect all broadcastable transactions.
+    pub fn script(&mut self, address: Address, calldata: Bytes) -> eyre::Result<ScriptResult> {
+        self.call(self.sender, address, calldata, U256::zero(), false)
+    }
+
+    /// Runs a broadcastable transaction locally and persists its state.
+    pub fn simulate(
+        &mut self,
+        from: Address,
+        to: Option<NameOrAddress>,
+        calldata: Option<Bytes>,
+        value: Option<U256>,
+    ) -> eyre::Result<ScriptResult> {
+        if let Some(NameOrAddress::Address(to)) = to {
+            self.call(from, to, calldata.unwrap_or_default(), value.unwrap_or(U256::zero()), true)
+        } else if to.is_none() {
+            let DeployResult { address: _, gas, logs, traces, debug } = self.executor.deploy(
+                from,
+                calldata.expect("No data for create transaction").0,
+                value.unwrap_or(U256::zero()),
+                None,
+            )?;
+
+            Ok(ScriptResult {
+                returned: bytes::Bytes::new(),
+                success: true,
+                gas,
+                logs,
+                traces: traces
+                    .map(|mut traces| {
+                        // Manually adjust gas for the trace to add back the stipend/real used gas
+                        traces.arena[0].trace.gas_cost = gas;
+                        vec![(TraceKind::Execution, traces)]
+                    })
+                    .unwrap_or_default(),
+                debug: vec![debug].into_iter().collect(),
+                labeled_addresses: Default::default(),
+                transactions: Default::default(),
+            })
+        } else {
+            eyre::bail!("ENS not supported.");
+        }
+    }
+
+    fn call(
+        &mut self,
+        from: Address,
+        to: Address,
+        calldata: Bytes,
+        value: U256,
+        commit: bool,
+    ) -> eyre::Result<ScriptResult> {
+        let RawCallResult {
+            result,
+            reverted,
+            gas: tx_gas,
+            stipend,
+            logs,
+            traces,
+            labels,
+            debug,
+            transactions,
+            ..
+        } = if !commit {
+            self.executor.call_raw(from, to, calldata.0, value)?
+        } else {
+            self.executor.call_raw_committing(from, to, calldata.0, value)?
+        };
+
+        let gas = if commit { tx_gas } else { tx_gas.overflowing_sub(stipend).0 };
```

Comment:
why do we subtract only if we're not committing?

Reply:
We usually only commit for simulating the actual `on-chain` transactions. 

The non-commit calls are reserved for the `run()` method that we call with `forge script`. And we usually (eg `forge run/test`) do not include the stipend on the traces.